lib/kafka_ex/server_kayrock.ex

```diff
@@ -0,0 +1,946 @@
+    end)
+  end
+
+  defp update_metadata(state, topics \\ []) do
+    # make sure we update metadata about known topics
+    known_topics = ClusterMetadata.known_topics(state.cluster_metadata)
+    topics = Enum.uniq(known_topics ++ topics)
+
+    {updated_state, response} =
+      retrieve_metadata(
+        state,
+        config_sync_timeout(),
+        topics
+      )
+
+    case response do
+      nil ->
+        updated_state
+
+      _ ->
+        new_cluster_metadata =
+          ClusterMetadata.from_metadata_v1_response(response)
+
+        {updated_cluster_metadata, brokers_to_close} =
+          ClusterMetadata.merge_brokers(
+            updated_state.cluster_metadata,
+            new_cluster_metadata
+          )
+
+        for broker <- brokers_to_close do
+          Logger.log(
+            :debug,
+            "Closing connection to broker #{broker.node_id}: #{
+              inspect(broker.host)
+            } on port #{inspect(broker.port)}"
+          )
+
+          NetworkClient.close_socket(broker.socket)
+        end
+
+        updated_state =
+          State.update_brokers(
+            %{updated_state | cluster_metadata: updated_cluster_metadata},
+            &maybe_connect_broker(&1, state)
+          )
+
+        updated_state
+    end
+  end
+
+  defp maybe_connect_broker(broker, state) do
+    case Broker.connected?(broker) do
+      true ->
+        broker
+
+      false ->
+        %{
+          broker
+          | socket:
+              NetworkClient.create_socket(
+                broker.host,
+                broker.port,
+                state.ssl_options,
+                state.use_ssl
+              )
+        }
+    end
+  end
+
+  defp retrieve_metadata(
+         state,
+         sync_timeout,
+         topics
+       ) do
+    retrieve_metadata(
+      state,
+      sync_timeout,
+      topics,
+      @retry_count,
+      0
+    )
+  end
+
+  defp retrieve_metadata(
+         state,
+         _sync_timeout,
+         topics,
+         0,
+         error_code
+       ) do
+    Logger.log(
+      :error,
+      "Metadata request for topics #{inspect(topics)} failed " <>
+        "with error_code #{inspect(error_code)}"
+    )
+
+    {state, nil}
+  end
+
+  defp retrieve_metadata(
+         state,
+         sync_timeout,
+         topics,
+         retry,
+         _error_code
+       ) do
+    # default to version 4 of the metdata protocol because this one treats an
+    # empty list of topics as 'no topics'.  note this limits us to kafka 0.11+
+    api_version = State.max_supported_api_version(state, :metadata, 4)
+
+    metadata_request = %{
+      Kayrock.Metadata.get_request_struct(api_version)
+      | topics: topics,
+        allow_auto_topic_creation: state.allow_auto_topic_creation
+    }
+
+    {{ok_or_err, response}, state_out} =
+      kayrock_network_request(metadata_request, :any, state)
+
+    case ok_or_err do
+      :ok ->
+        # THIS WILL PROBABLY NOT WORK
```

Comment:
Should we be worried ? ^_^

Reply:
Haha some of this is comments left over from me transposing the old code.  My guess is I just forgot to delete the comments - I'll do some double checking.

Reply:
Ya I checked with the coverage tool, it works.  Just an old comment.